service/header/sync.go

```diff
@@ -254,40 +301,88 @@ func (s *Syncer) incoming(ctx context.Context, maybeHead *ExtendedHeader) pubsub
 var requestSize uint64 = 512
 
 // syncTo requests headers from locally stored head up to the new head.
-func (s *Syncer) syncTo(ctx context.Context, newHead *ExtendedHeader) error {
+func (s *Syncer) syncTo(ctx context.Context, newHead *ExtendedHeader) {
 	head, err := s.store.Head(ctx)
 	if err != nil {
-		return err
+		log.Errorw("getting head during sync", "err", err)
+		return
 	}
 
 	if head.Height == newHead.Height {
-		return nil
+		return
+	}
+
+	log.Infow("syncing headers",
+		"from", head.Height,
+		"to", newHead.Height)
+	err = s.doSync(ctx, head, newHead)
+	if err != nil {
+		log.Errorw("syncing headers",
+			"from", head.Height,
+			"to", newHead.Height,
+			"err", err)
+		return
 	}
 
-	log.Infow("syncing headers", "from", head.Height, "to", newHead.Height)
-	defer log.Info("synced headers")
+	log.Infow("synced headers",
+		"from", head.Height,
+		"to", newHead.Height,
+		"took", s.state.End.Sub(s.state.Start))
+}
 
-	start, end := uint64(head.Height)+1, uint64(newHead.Height)
-	for start <= end {
-		amount := end - start + 1
+// doSync performs actual syncing updating the internal SyncState
+func (s *Syncer) doSync(ctx context.Context, oldHead, newHead *ExtendedHeader) (err error) {
+	from, to := uint64(oldHead.Height)+1, uint64(newHead.Height)
+
+	s.stateLk.Lock()
+	s.state.ID++
+	s.state.FromHeight = from
+	s.state.ToHeight = to
+	s.state.FromHash = oldHead.Hash()
+	s.state.ToHash = newHead.Hash()
+	s.state.Start = time.Now()
+	s.stateLk.Unlock()
+
+	for from <= to {
+		amount := to - from + 1
 		if amount > requestSize {
 			amount = requestSize
 		}
 
-		headers, err := s.getHeaders(ctx, start, amount)
-		if err != nil && len(headers) == 0 {
-			return err
+		amount, err = s.processHeaders(ctx, from, amount)
+		if err != nil && amount == 0 {
+			break
 		}
 
-		err = s.store.Append(ctx, headers...)
-		if err != nil {
-			return err
-		}
+		from += amount
+		s.stateLk.Lock()
+		s.state.Height = from
+		s.stateLk.Unlock()
+	}
+
+	s.stateLk.Lock()
+	s.state.Height = from - 1 // minus one as we add one to the amount above
```

Comment:
I would prefer to use a new variable instead of clobbering `from` above then having to go back and fix it. The way it's currently done _requires_ a comment, but using a new variable would be self-documenting.

Reply:
Good suggestion, even though I prefer less code than more. Also, I will still preserve the comment, just in case.